cdap-api/src/main/java/co/cask/cdap/api/TransactionUtil.java

```diff
@@ -0,0 +1,215 @@
+      throw propagate(e);
+    }
+    return result.get();
+  }
+
+  /**
+   * Executes the given {@link TxCallable} using the given {@link Transactional}.
+   *
+   * @param transactional the {@link Transactional} to use for transactional execution.
+   * @param callable the {@link TxCallable} to be executed inside a transaction
+   * @param <V> type of the result
+   * @return value returned by the given {@link TxCallable}.
+   * @throws X if failed to execute the given {@link TxRunnable} in a transaction. If the TransactionFailureException
+   * has a cause in it, the cause is thrown as-is if it is an instance of X.
+   */
+  public static <V, X extends Throwable> V execute(Transactional transactional,
+                              final TxCallable<V> callable, Class<X> exception) throws X {
+    final AtomicReference<V> result = new AtomicReference<>();
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          result.set(callable.call(context));
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw propagate(e, exception);
+    }
+    return result.get();
+  }
+
+  /**
+   * Propagates {@code throwable} as-is if it is an instance of
+   * {@link RuntimeException} or {@link Error}, or else as a last resort, wraps
+   * it in a {@code RuntimeException} then propagates.
+   * @param throwable the Throwable to propagate
+   * @return nothing will ever be returned; this return type is only for convenience
+   */
+  public static RuntimeException propagate(Throwable throwable) {
+    propagateIfPossible(checkNotNull(throwable));
+    throw new RuntimeException(throwable);
+  }
+
+  /**
+   * Propagates the given {@link TransactionFailureException}. If the {@link TransactionFailureException#getCause()}
+   * doesn't return {@code null}, the cause will be used instead for the propagation. This method will
+   * throw the failure exception as-is the given propagated type if the type matches or as {@link RuntimeException}.
+   * This method will always throw exception and the returned exception is for satisfying Java static analysis only.
+   *
+   * @param e the {@link TransactionFailureException} to propagate
+   * @param propagateType if the exception is an instance of this type, it will be rethrown as is
+   * @param <X> exception type of propagate type
+   * @return a exception of type X.
+   */
+  public static <X extends Throwable> X propagate(TransactionFailureException e,
+                                                  Class<X> propagateType) throws X {
+    Throwable cause = firstNonNull(e.getCause(), e);
+    propagateIfPossible(cause, propagateType);
+    throw propagate(cause);
+  }
+
+  /**
+   * Propagates {@code throwable} exactly as-is, if and only if it is an
+   * instance of {@link RuntimeException} or {@link Error}.
+   */
+  public static void propagateIfPossible(@Nullable Throwable throwable) {
```

Comment:
Make all the `propagateIfPossible` to private method and take `TransactionFailureException` only.